### PR TITLE
Implement SPI driver on ATmega2560

### PR DESCRIPTION
**Pull request by danielamkaer**

### CODE REVIEW

**Comment by gebart** [comment]
@danielamkaer Please add a prefix to your commit message, like "cpu/atmega2560: Add SPI driver implementation", and also squash your whitespace fixes into the main commit.

I don't have the hardware available right now so I can't test. @N8Fear or @LudwigOrtmann, do you have any SPI devices that you can wire up to the Arduino?

File: boards/arduino-mega2560/include/periph_conf.h [hunk 1]
```diff
@@ -250,10 +250,28 @@ extern "C" {
 /**
  * @brief SPI configuration
  */
-#define SPI_NUMOF           (0U)                                                    /* TODO */
-#define SPI_0_EN            0
+#define SPI_NUMOF           (1U)
+#define SPI_0_EN            1
 #define SPI_1_EN            0
 
+// SPI 0 Registers
```
Comment (thomaseichinger) [inline boards/arduino-mega2560/include/periph_conf.h]: Please use C comment style. `//` -> `/*  */`

File: boards/arduino-mega2560/include/periph_conf.h [hunk 2]
```diff
@@ -250,10 +250,28 @@ extern "C" {
...
+#define SPI_0_CR 			SPCR
+#define SPI_0_DR 			SPDR
+#define SPI_0_SR 			SPSR
+#define SPI_0_DDR 			DDRB
+#define SPI_0_PORT 			PORTB
```
Comment (thomaseichinger) [inline boards/arduino-mega2560/include/periph_conf.h]: `SPI_0_PORT` is not used, could you remove it?

File: cpu/atmega2560/periph/spi.c [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+#if SPI_1_EN
+    [SPI_1] = MUTEX_INIT,
+#endif
+};
+
+int spi_init_master(spi_t dev, spi_conf_t conf, spi_speed_t speed)
+{
+    uint8_t _spcr = (1<<SPI_0_MSTR)|(1<<SPI_0_SPE);
+
+    // Assume F_CPU = 16.000.000 Hz
```
Comment (thomaseichinger) [inline cpu/atmega2560/periph/spi.c]: Here and all following, please change `//` -> `/*  */`

File: cpu/atmega2560/periph/spi.c [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+            // Actual rate = 250.000 Hz
+            _spcr |= (1<<SPI_0_SPR1);
+            break;
+        case SPI_SPEED_1MHZ:
+            // Actual rate = 1.000.000 Hz
+            _spcr |= (1<<SPI_0_SPR0);
+            break;
+        case SPI_SPEED_5MHZ:
+            // Actual rate = 4.000.000 Hz
+            _spcr |= 0;
```
Comment (thomaseichinger) [inline cpu/atmega2560/periph/spi.c]: Line without effect? Do you mind removing it and add a comment stating 4MHz is default?

File: cpu/atmega2560/periph/spi.c [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        case SPI_CONF_SECOND_FALLING:
+            _spcr |= (1<<SPI_0_CPOL)|(1<<SPI_0_CPHA);
+            break;
+    }
+
+    switch (dev) {
+#if SPI_0_EN
+        case SPI_0:
+            SPI_0_DDR |= (1<<SPI_0_MOSI)|(1<<SPI_0_SCK);
+            SPI_0_DDR &= ~((1<<SPI_0_MISO));
```
Comment (thomaseichinger) [inline cpu/atmega2560/periph/spi.c]: These two lines could be moved to `spi_conf_pins`?

File: cpu/atmega2560/periph/spi.c [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+        case SPI_CONF_SECOND_FALLING:
+            _spcr |= (1<<SPI_0_CPOL)|(1<<SPI_0_CPHA);
+            break;
+    }
+
+    switch(dev) {
+#if SPI_0_EN
+        case SPI_0:
+            SPI_0_DDR |= (1<<SPI_0_MISO);
+            SPI_0_DDR &= ~((1<<SPI_0_MOSI)|(SPI_0_SCK));
```
Comment (thomaseichinger) [inline cpu/atmega2560/periph/spi.c]: Also, `spi_conf_pins`?

File: cpu/atmega2560/periph/spi.c [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+    }
+    mutex_unlock(&locks[dev]);
+    return 0;
+}
+
+int spi_transfer_byte(spi_t dev, char out, char *in) {
+    switch (dev) {
+#if SPI_0_EN
+    case SPI_0:
+        SPI_0_DR = out;
```
Comment (thomaseichinger) [inline cpu/atmega2560/periph/spi.c]: I'm not a ATmega expert, but wouldn't it be better to check if a previously set data register was already written out?

File: cpu/atmega2560/periph/spi.c [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+        break;
+#endif
+    }
+    return -1;
+}
+
+int spi_transfer_bytes(spi_t dev, char *out, char *in, unsigned int length) {
+    unsigned int i = 0;
+    for (i=0; i < length; i++) {
+        int res = spi_transfer_byte(dev, (out != NULL?out[i]:0),
```
Comment (thomaseichinger) [inline cpu/atmega2560/periph/spi.c]: Do you mind spending these `?:`s some spaces?

File: cpu/atmega2560/periph/spi.c (hunk 5 above)
Comment (danielamkaer) [inline cpu/atmega2560/periph/spi.c]: The configuration of the pins depends on whether it is master/slave, no parameter for spi_conf_pins to configure that

File: cpu/atmega2560/periph/spi.c (hunk 8 above)
Comment (thomaseichinger) [inline cpu/atmega2560/periph/spi.c]: And maybe some braces? `((out != NULL) ? out[i] : 0)`

File: cpu/atmega2560/periph/spi.c [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+        res = spi_transfer_byte(dev,(out != NULL ? out[i] : 0),
+                (in != NULL?&(in[i+1]):NULL));
+        if (res < 0) {
+            return res;
+        }
+    }
+    return i;
+}
+
+// UNSURE: void spi_transmission_begin(spi_t dev, char reset_val);
```
Comment (thomaseichinger) [inline cpu/atmega2560/periph/spi.c]: What are you unsure about?

File: cpu/atmega2560/periph/spi.c (hunk 7 above)
Comment (danielamkaer) [inline cpu/atmega2560/periph/spi.c]: As far as I can read from the data sheet, there isn't a way to check if the outgoing data register is empty. However the function will not return until SPIF is set, which indicates that the operation is finished.

File: cpu/atmega2560/periph/spi.c (hunk 5 above)
Comment (thomaseichinger) [inline cpu/atmega2560/periph/spi.c]: Ah, didn't see `SPI_0_SCK` changed lines.

**Comment by OlegHahm** [comment]
Welcome to RIOT, @danielamkaer!

**Comment by thomaseichinger** [comment]
@danielamkaer Great job! Please see some minor remarks above.

File: cpu/atmega2560/periph/spi.c (hunk 9 above)
Comment (danielamkaer) [inline cpu/atmega2560/periph/spi.c]: Exactly what spi_transmission_begin should do?

**Comment by LudwigOrtmann** [comment]
I have, but I probably won't get around to testing this PR in the immediate future.

**Comment by PeterKietzmann** [comment]
Now that I gave up my Atmega2560 board, testing this PR might also be a job for @ReneHerthel :-)

**Comment by LudwigKnuepfer** [comment]
With
```diff
diff --git a/boards/arduino-mega2560/Makefile.features b/boards/arduino-mega2560/Makefile.features
index 6d25415..ad1a466 100644
--- a/boards/arduino-mega2560/Makefile.features
+++ b/boards/arduino-mega2560/Makefile.features
@@ -1,4 +1,5 @@
 FEATURES_PROVIDED += periph_uart
 FEATURES_PROVIDED += periph_gpio
 FEATURES_PROVIDED += periph_timer
+FEATURES_PROVIDED += periph_spi
 FEATURES_MCU_GROUP = avr8
 ```
I get
```
Building application "periph_spi" for "arduino-mega2560" with MCU "atmega2560".

"make" -C /home/lo/RIOT/boards/arduino-mega2560
"make" -C /home/lo/RIOT/core
"make" -C /home/lo/RIOT/cpu/atmega2560
"make" -C /home/lo/RIOT/cpu/atmega2560/periph
"make" -C /home/lo/RIOT/cpu/atmega_common
"make" -C /home/lo/RIOT/drivers
"make" -C /home/lo/RIOT/sys
"make" -C /home/lo/RIOT/sys/auto_init
"make" -C /home/lo/RIOT/sys/shell
"make" -C /home/lo/RIOT/sys/shell/commands
"make" -C /home/lo/RIOT/sys/tsrb
"make" -C /home/lo/RIOT/sys/uart_stdio
/home/lo/RIOT/tests/periph_spi/bin/arduino-mega2560/periph_spi.a(main.o): In function `slave_on_cs':
/home/lo/RIOT/tests/periph_spi/main.c:135: undefined reference to `spi_transmission_begin'
collect2: error: ld returned 1 exit status
/home/lo/RIOT/tests/periph_spi/../../Makefile.include:213: recipe for target 'all' failed
make: *** [all] Error 1
```
Please ping me when this is resolved.

**Comment by danielamkaer** [comment]
@LudwigKnuepfer try now.

**Comment by LudwigKnuepfer** [comment]
@danielamkaer on a general note regarding our development procedures - please don't instantly squash your commits, it makes it easier for reviewers to see what changed. The "needs squashing" label is not there to alert you, but to prevent merging of unsquashed PRs.

**Comment by danielamkaer** [comment]
Okay @LudwigKnuepfer - I'll remember for the future. This PR + 2 others are my first ever contributions to open source projects.

**Comment by LudwigKnuepfer** [comment]
@danielamkaer No sweat, you'll (hopefully) get more feedback like this during your first steps. And remember: although some work has gone into our processes and policies, they are not perfect and can benefit from your input. So don't refrain from speaking out if you see room for improvement.

**Comment by LudwigKnuepfer** [comment]
Can't test, don't know how to map GPIO pins to board layout anymore due to changes in RIOT =(
@haukepetersen I suspect you had your hands in this. Please tutor me ;)

**Comment by danielamkaer** [comment]
@LudwigKnuepfer - Lookup using https://www.arduino.cc/en/Hacking/PinMapping2560

Translate pin 4 (PG5) on Arduino into ```GPIO_PIN(PORT_G, 5)```

**Comment by danielamkaer** [comment]
Any progress?

**Comment by haukepetersen** [comment]
@LudwigKnuepfer I guess we can shed some light on the pin mapping: we don't have a sub-set statically mapped anymore, but you can access any MCU pin freely using the `GPIO_PIN(port, pin)` macro. As @danielamkaer got it right, if you e.g. want to set pin PG.05 you just call `gpio_set(GPIO_PIN(PORT_G, 5)` or `gpio_set(GPIO(5, 5))`. The port enum and the `GPIO_PIN` macros are defined in `cpu/atmega2560/cpu_periph.h`.

Now mapping the acutal MCU pins to Arduino pins is another issue, just refer to the Arduino doc (e.g. the link posted above) for details. But I have an idea how to make this easier - will prepare a PR...

**Comment by haukepetersen** [comment]
how about this: #4313. Now you can just use `gpio_xx(ARDUINO_PIN_X)`

**Comment by LudwigKnuepfer** [comment]
Sorry, I did not have time for testing.. I would not feel hurt if anyone else tests this ;)

**Comment by zhuoshuguo** [comment]
@danielamkaer Ping. A kind reminder for rebase.

**Comment by danielamkaer** [comment]
@zhuoshuguo I need some help I think. I'm trying to do
$ git pull upstream master --rebase
but I run into a lot of conflicts, which are not caused by my changes.

**Comment by LudwigKnuepfer** [comment]
@danielamkaer 
I just did a local `git rebase upstream/master` with your branch and that went fine.

**Comment by OlegHahm** [comment]
@thomaseichinger, @haukepetersen, @PeterKietzmann, anyone for a review?

**Comment by PeterKietzmann** [comment]
Yes, in general but not so much time currently. At the latest during the next Hack'n'ACK

**Comment by kYc0o** [comment]
Maybe @aabadie would like to try this? (:

**Comment by aabadie** [comment]
Thanks @kYc0o. Why not, if I have time.

**Comment by haukepetersen** [comment]
Hi, just tested this PR, for me it does not work. Connecting a logic analyzer to pins 50-52 does not show me any pin activity when sending data on the bus...

**Comment by haukepetersen** [comment]
While starting to debug the driver, I did so many changes, that I decided to make a new PR -> #5020.

@danielamkaer: Sorry again for starting to actually test this so late! It's the good old thing that you only start with something until you really need it...

**Comment by haukepetersen** [comment]
closed in favor of #5020

**Pull request closed without merging**